### DISCUSSION TITLE
How can I write username and login time into logfile when user log in?

### DISCUSSION THREAD

**Discussion by skk294** [fastapi-users/fastapi-users#683] [Q&A]
I have tried middleware but only got time there. If I read the body from request.body the response won't be sent. I don't know how to do it now.

**Comment by frankie567** [comment]
Hi there,

Unfortunately, we don't provide an `on_after_login` callback, which should be the easiest way to do.

That said, going with a middleware should work quite easily. My guess is that you forgot to call `call_next` in your implementation which should trigger the actual route. You should have a look at FastAPI documentation: https://fastapi.tiangolo.com/tutorial/middleware/?h=middle#create-a-middleware

**Reply by skk294** [reply]
Thanks for answering my question. What I mean is once I comsume the request.body, I dont know how to restore it so that causes the problem. Now I use middleware to catch 'fastapiusersauth' in response.header, then use jwt.decode to get user_id, it works.
I hope there will be an on_after_login callback.

**Reply by frankie567** [reply]
I see. Have you tried to use `request.form()` instead? It gives you an already parsed representation of the body which should not be sensible to the "already consumed" issue.

**Reply by skk294** [reply]
I have tried, `request.form()` is a `<coroutine object Request.form at 0x7f01a166dc40>`, have to use `await request.form()` to get content.Then, same problem as use `await request.body()`

**Comment by bkis** [comment]
By now there's an [`on_after_login` callback for the UserManager](https://fastapi-users.github.io/fastapi-users/11.0/configuration/user-manager/#on_after_login).
